operator/auth/spire/client.go

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package spire
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	entryv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/entry/v1"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/cilium/cilium/operator/auth/identity"
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+const (
+	notFoundError   = "NotFound"
+	defaultParentID = "/dclient"
+)
+
+var defaultSelectors = []*types.Selector{
+	{
+		Type:  "cilium",
+		Value: "mtls",
+	},
+}
+
+// Cell is the cell for the SPIRE client.
+var Cell = cell.Module(
+	"spire-client",
+	"Spire Server API Client",
+	cell.Config(ClientConfig{}),
+	cell.Provide(NewClient),
+)
+
+// ClientConfig contains the configuration for the SPIRE client.
+type ClientConfig struct {
+	AuthMTLSEnabled      bool   `mapstructure:"mesh-auth-mtls-enabled"`
+	SpireAgentSocketPath string `mapstructure:"mesh-auth-spire-agent-socket"`
+	SpireServerAddress   string `mapstructure:"mesh-auth-spire-server-address"`
+	SpiffeTrustDomain    string `mapstructure:"mesh-auth-spiffe-trust-domain"`
+}
+
+// Flags adds the flags used by ClientConfig.
+func (cfg ClientConfig) Flags(flags *pflag.FlagSet) {
+	flags.BoolVar(&cfg.AuthMTLSEnabled,
+		"mesh-auth-mtls-enabled",
+		false,
+		"The flag to enable mTLS for the SPIRE server.")
+	flags.StringVar(&cfg.SpireAgentSocketPath,
+		"mesh-auth-spire-agent-socket",
+		"/run/spire/sockets/agent/agent.sock",
+		"The path for the SPIRE admin agent Unix socket.")
+	flags.StringVar(&cfg.SpireServerAddress,
+		"mesh-auth-spire-server-address",
+		"spire-server.spire.svc.cluster.local:8081",
+		"SPIRE server endpoint.")
+	flags.StringVar(&cfg.SpiffeTrustDomain,
+		"mesh-auth-spiffe-trust-domain",
+		"spiffe.cilium.io",
+		"The trust domain for the SPIFFE identity.")
+}
+
+type Client struct {
+	cfg   ClientConfig
+	entry entryv1.EntryClient
+}
+
+// NewClient creates a new SPIRE client.
+// If the mTLS is not enabled, it returns a noop client.
+func NewClient(cfg ClientConfig) (identity.Provider, error) {
+	if !cfg.AuthMTLSEnabled {
+		return &noopClient{}, nil
+	}
+
+	// This is blocking till the cilium-operator is registered in SPIRE.
+	source, err := workloadapi.NewX509Source(
+		context.Background(),
+		workloadapi.WithClientOptions(
+			workloadapi.WithAddr(fmt.Sprintf("unix://%s", cfg.SpireAgentSocketPath)),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())
```

Comment:
for added security we could check on the SVID that the admin api offers us to confirm we are indeed talking to the right one. There is a built in function for that in the spiffe client lib

Reply:
good point, for the time being, I think we can just verify based on the trusted domain only. More granular checks (with SPIFEE id) can be done once we have done all configuration/setup as part of SPIFEE installation.